src/capi_frontend/capi_utils.cpp

```diff
@@ -0,0 +1,142 @@
+#include "../precision.hpp"
+#include "../profiler.hpp"
+#include "../status.hpp"
+
+namespace ovms {
+Precision PrecisionToOvmsPrecision(const OVMSDataType& datatype) {
+    static std::unordered_map<OVMSDataType, Precision> precisionMap{
+        {OVMS_DATATYPE_BOOL, Precision::BOOL},
+        {OVMS_DATATYPE_FP64, Precision::FP64},
+        {OVMS_DATATYPE_FP32, Precision::FP32},
+        {OVMS_DATATYPE_FP16, Precision::FP16},
+        {OVMS_DATATYPE_I64, Precision::I64},
+        {OVMS_DATATYPE_I32, Precision::I32},
+        {OVMS_DATATYPE_I16, Precision::I16},
+        {OVMS_DATATYPE_I8, Precision::I8},
+        {OVMS_DATATYPE_U64, Precision::U64},
+        {OVMS_DATATYPE_U32, Precision::U32},
+        {OVMS_DATATYPE_U16, Precision::U16},
+        // {"BYTES", Precision::??},
+        {OVMS_DATATYPE_U8, Precision::U8}};
+    auto it = precisionMap.find(datatype);
+    if (it == precisionMap.end()) {
+        return Precision::UNDEFINED;
+    }
+    return it->second;
+}
+
+size_t DataTypeSize(const OVMSDataType& datatype) {
+    static std::unordered_map<OVMSDataType, size_t> datatypeSizeMap{
+        {OVMS_DATATYPE_BOOL, 1},
+        {OVMS_DATATYPE_U8, 1},
+        {OVMS_DATATYPE_U16, 2},
+        {OVMS_DATATYPE_U32, 4},
+        {OVMS_DATATYPE_U64, 8},
+        {OVMS_DATATYPE_I8, 1},
+        {OVMS_DATATYPE_I16, 2},
+        {OVMS_DATATYPE_I32, 4},
+        {OVMS_DATATYPE_I64, 8},
+        {OVMS_DATATYPE_FP16, 2},
+        {OVMS_DATATYPE_FP32, 4},
+        {OVMS_DATATYPE_FP64, 8}
+        // {"BYTES", },
+    };
+    auto it = datatypeSizeMap.find(datatype);
+    if (it == datatypeSizeMap.end()) {
+        return 0;
+    }
+    return it->second;
+}
+
+const OVMSDataType& ovmsPrecisionToPrecision(Precision precision) {
+    static std::unordered_map<Precision, OVMSDataType> precisionMap{
+        {Precision::FP64, OVMS_DATATYPE_FP64},
+        {Precision::FP32, OVMS_DATATYPE_FP32},
+        {Precision::FP16, OVMS_DATATYPE_FP16},
+        {Precision::I64, OVMS_DATATYPE_I64},
+        {Precision::I32, OVMS_DATATYPE_I32},
+        {Precision::I16, OVMS_DATATYPE_I16},
+        {Precision::I8, OVMS_DATATYPE_I8},
+        {Precision::U64, OVMS_DATATYPE_U64},
+        {Precision::U32, OVMS_DATATYPE_U32},
+        {Precision::U16, OVMS_DATATYPE_U16},
+        {Precision::U8, OVMS_DATATYPE_U8},
+        {Precision::BOOL, OVMS_DATATYPE_BOOL},
+        {Precision::INVALID, OVMS_DATATYPE_INVALID}};
+
+    // {Precision::BF16, ""},
+    // {Precision::U4, ""},
+    // {Precision::U1, ""},
+    // {Precision::CUSTOM, ""},
+    // {Precision::DYNAMIC, ""},
+    // {Precision::MIXED, ""},
+    // {Precision::Q78, ""},
+    // {Precision::BIN, ""},
+    // {Precision::I4, ""},
+    // {Precision::UNDEFINED, "UNDEFINED"}};
+
+    auto it = precisionMap.find(precision);
+    if (it == precisionMap.end()) {
+        it = precisionMap.find(Precision::INVALID);
+    }
+    return it->second;
+}
+
+std::string tensorShapeToString(const Shape& shape) {
```

Comment:
This is not related to C-API but inside capi_utils.cpp

Reply:
We can refactor this later. Currently implemented as in other APIs.

Reply:
Use Shape::toString()